### PR TITLE
Feature/panda 3209 UT for customer coupon

### PR DESCRIPTION
**Pull request by Jerify02**

### CODE REVIEW

File: projects/storefrontlib/src/cms-components/myaccount/my-coupons/coupon-card/coupon-card.component.spec.ts [hunk 1]
```diff
@@ -39,6 +40,7 @@ describe('CouponCardComponent', () => {
 
   beforeEach(() => {
     fixture = TestBed.createComponent(CouponCardComponent);
+    el = fixture.debugElement;
     component = fixture.componentInstance;
     component.coupon = mockCoupon;
     component.couponLoading = false;
```
Comment (kate1zheng) [inline projects/storefrontlib/src/cms-components/myaccount/my-coupons/coupon-card/coupon-card.component.spec.ts]: input/output - can we using a test host to simulate the real usage https://angular.io/guide/testing#component-inside-a-test-host

File: projects/storefrontlib/src/cms-components/myaccount/my-coupons/coupon-card/coupon-card.component.spec.ts [hunk 2]
```diff
@@ -50,82 +52,36 @@ describe('CouponCardComponent', () => {
...
     fixture.detectChanges();
-    const readMoreLink = fixture.debugElement.query(By.css('a'));
-    readMoreLink.nativeElement.click();
+    el.query(By.css('[data-test]="read-more"')).nativeElement.click();
     expect(modalService.open).toHaveBeenCalled();
   });
 
-  it('should be able to subscribe/unsubscribe coupon notification', () => {
+  it('should be able to subscribe/unsubscribe notification', () => {
     spyOn(component.notificationChanged, 'emit').and.callThrough();
```
Comment (kate1zheng) [inline projects/storefrontlib/src/cms-components/myaccount/my-coupons/coupon-card/coupon-card.component.spec.ts]: it seems that this line is not needed

File: projects/storefrontlib/src/cms-components/myaccount/my-coupons/coupon-claim/coupon-claim.component.spec.ts [hunk 3]
```diff
@@ -1,5 +1,5 @@
 import { async, ComponentFixture, TestBed } from '@angular/core/testing';
-import { CouponClaimComponent } from './coupon-claim.component'
+import { CouponClaimComponent } from './coupon-claim.component';
 import {
   UserService,
   RoutingService,
```
Comment (kate1zheng) [inline projects/storefrontlib/src/cms-components/myaccount/my-coupons/coupon-claim/coupon-claim.component.spec.ts]: please only include 1 describe in spec

Comment (kate1zheng) [reply projects/storefrontlib/src/cms-components/myaccount/my-coupons/coupon-claim/coupon-claim.component.spec.ts]: no need to call ngOnInit manually

File: projects/storefrontlib/src/cms-components/myaccount/my-coupons/my-coupons.component.spec.ts [hunk 4]
```diff
@@ -114,69 +117,45 @@ describe('MyCouponsComponent', () => {
...
+  it('should be able to change page/sort', () => {
     fixture.detectChanges();
-    expect(userService.loadCustomerCoupons).toHaveBeenCalledWith(
-      10,
-      1,
-      'startDate:asc'
-    );
+
+    component.sortChange('sort');
+    expect(userService.loadCustomerCoupons).toHaveBeenCalled();
```
Comment (kate1zheng) [inline projects/storefrontlib/src/cms-components/myaccount/my-coupons/my-coupons.component.spec.ts]: please add parameter

File: projects/storefrontlib/src/cms-components/myaccount/my-coupons/my-coupons.component.spec.ts [hunk 5]
```diff
@@ -114,69 +117,45 @@ describe('MyCouponsComponent', () => {
...
+
+    component.pageChange(2);
+    expect(userService.loadCustomerCoupons).toHaveBeenCalled();
```
Comment (kate1zheng) [inline projects/storefrontlib/src/cms-components/myaccount/my-coupons/my-coupons.component.spec.ts]: add parameter

File: projects/storefrontlib/src/cms-components/myaccount/my-coupons/coupon-card/coupon-card.component.spec.ts [hunk 6]
```diff
@@ -25,107 +31,85 @@ class MockUrlPipe implements PipeTransform {
...
+    expect(el.query(By.css('[data-test]="notify-checkbox"'))).toBeTruthy();
+    expect(el.query(By.css('[data-test]="find-button"'))).toBeTruthy();
   });
 
-  it('should be able to open coupon detail dialog', () => {
+  it('should be able to open "read more" dialog', () => {
     fixture.detectChanges();
-    const readMoreLink = fixture.debugElement.query(By.css('a'));
-    readMoreLink.nativeElement.click();
+    el.query(By.css('.cx-card-read-more')).nativeElement.click();
```
Comment (Qingyu-Yang) [inline projects/storefrontlib/src/cms-components/myaccount/my-coupons/coupon-card/coupon-card.component.spec.ts]: use data-test

File: projects/storefrontlib/src/cms-components/myaccount/my-coupons/coupon-card/coupon-card.component.spec.ts [hunk 7]
```diff
@@ -25,107 +31,85 @@ class MockUrlPipe implements PipeTransform {
...
 
-  it('should be able to subscribe/unsubscribe coupon notification', () => {
-    spyOn(component.notificationChanged, 'emit').and.callThrough();
+  it('should be able to subscribe/unsubscribe notification', () => {
     fixture.detectChanges();
-    const couponNotificationCheckbox = fixture.debugElement.query(
-      By.css('.form-check-input')
-    );
-    couponNotificationCheckbox.nativeElement.click();
+    const chx = el.query(By.css('.form-check-input')).nativeElement;
```
Comment (Qingyu-Yang) [inline projects/storefrontlib/src/cms-components/myaccount/my-coupons/coupon-card/coupon-card.component.spec.ts]: use data-test

File: projects/storefrontlib/src/cms-components/myaccount/my-coupons/coupon-card/coupon-card.component.spec.ts [hunk 8]
```diff
@@ -25,107 +31,85 @@ class MockUrlPipe implements PipeTransform {
...
-    expect(couponNotificationCheckbox[0].nativeElement.checked).toBeFalsy();
-    component.coupon.notificationOn = true;
-    fixture.detectChanges();
-    couponNotificationCheckbox = fixture.debugElement.queryAll(
-      By.css('.form-check-input')
-    );
-    expect(couponNotificationCheckbox.length).toBe(1);
-    expect(couponNotificationCheckbox[0].nativeElement.checked).toBeTruthy();
+    chx.click();
+    expect(component.loading).toEqual(true);
```
Comment (Qingyu-Yang) [inline projects/storefrontlib/src/cms-components/myaccount/my-coupons/coupon-card/coupon-card.component.spec.ts]: Is it possible to  expect component.card.notificationChanged.emit.toHaveBeenCalledWith(xxx)?

Comment (Jerify02) [reply projects/storefrontlib/src/cms-components/myaccount/my-coupons/coupon-card/coupon-card.component.spec.ts]: component.loading==true indicates that event has been emitted.

**Pull request closed without merging**